lib/mlir/Dialect/Clift/Utils/ImportModel.cpp

```diff
@@ -0,0 +1,589 @@
+      RequireComplete = true;
+      const auto &T = llvm::cast<model::CABIFunctionType>(ModelType);
+      return getTypeAttribute(T);
+    }
+    case model::TypeKind::EnumType: {
+      RequireComplete = true;
+      const auto &T = llvm::cast<model::EnumType>(ModelType);
+      return getTypeAttribute(T);
+    }
+    case model::TypeKind::RawFunctionType: {
+      RequireComplete = true;
+      const auto &T = llvm::cast<model::RawFunctionType>(ModelType);
+      return getTypeAttribute(T);
+    }
+    case model::TypeKind::StructType: {
+      const auto &T = llvm::cast<model::StructType>(ModelType);
+      return getTypeAttribute(T, RequireComplete);
+    }
+    case model::TypeKind::TypedefType: {
+      const auto &T = llvm::cast<model::TypedefType>(ModelType);
+      return getTypeAttribute(T, RequireComplete);
+    }
+    case model::TypeKind::UnionType: {
+      const auto &T = llvm::cast<model::UnionType>(ModelType);
+      return getTypeAttribute(T, RequireComplete);
+    }
+
+    case model::TypeKind::PrimitiveType:
+      revng_abort("Primitive types have no corresponding attribute.");
+
+    case model::TypeKind::Invalid:
+    case model::TypeKind::Count:
+      revng_abort("These are invalid values. Something has gone wrong.");
+    }
+  }
+
+  RecursiveCoroutine<clift::ValueType>
+  getUnwrappedValueType(const model::Type &ModelType,
+                        bool RequireComplete = false,
+                        const bool Const = false) {
+    if (const auto *const T = llvm::dyn_cast<model::PrimitiveType>(&ModelType))
+      rc_return getPrimitiveType(*T, Const);
+
+    const auto getDefinedType = [&](const auto Attr) -> clift::ValueType {
+      return make<clift::DefinedType>(Attr, getBool(Const));
+    };
+
+    if (const auto It = Cache.find(ModelType.ID()); It != Cache.end())
+      rc_return getDefinedType(It->second);
+
+    const clift::TypeDefinition Attr = getTypeAttribute(ModelType,
+                                                        RequireComplete);
+
+    if (not Attr)
+      rc_return nullptr;
+
+    if (RequireComplete) {
+      const auto R = Cache.try_emplace(ModelType.ID(), Attr);
+      revng_assert(R.second);
+    }
+
+    rc_return getDefinedType(Attr);
+  }
+
+  RecursiveCoroutine<clift::ValueType>
+  getQualifiedValueType(const model::QualifiedType &ModelType,
+                        bool RequireComplete = false) {
+    if (not ModelType.UnqualifiedType().isValid()) {
+      if (EmitError)
+        EmitError() << "Invalid UnqualifiedType in QualifiedType";
+      rc_return nullptr;
+    }
+
+    // Qualifier::isConst verifies itself, which interferes with unit testing.
+    const auto IsConst = [](const model::Qualifier &Q) {
+      return Q.Kind() == model::QualifierKind::Const;
+    };
+
+    // Qualifier::isPointer verifies itself, which interferes with unit testing.
+    const auto IsPointer = [](const model::Qualifier &Q) {
+      return Q.Kind() == model::QualifierKind::Pointer;
+    };
```

Comment:
What do you mean they interfere with unit testing?

Reply:
They contain assertions verifying the type and some of the unit tests are testing for import failure. Using those functions would cause the assertions to fail and in turn cause the unit test to fail.

Reply:
Aaah no ok, I finally got what you meant here! It's ok then!